### PR TITLE
[PM-5533] Migrate Org Keys to state providers

### PR DESCRIPTION
**Pull request by MGibson1**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Migrates org keys to state providers and derived state. For this I adopted a pattern to extract key/derive definition to a separate file so that I can test deserializers and derive methods. Eventually I'd like to limit export of these definitions, but didn't want to move cryptoService and update imports all over the place.

## Code changes

- **encrypte-organization-key/guid.d.ts**: Added an `OrgKey` opaque type
- **crypto service abstraction**: adds org key observable, deprecates old get methods
  - this also updates the return type to a `Record` since there's no reason to use the `Map` and it's worse for serialization
  - **cipher.service**: This is collateral damage to the change from `Map` to `Record`
- **StateService / Account**: Remove org Keys from state service
- **org-keys**: File containing the definitions for state for organization keys
- **user-keys**: file containing the definitions for the state for user keys. This is just a move from crypto-service and will eventually contain more definitions for user keys
- **crypto service implementation**: adds state for encrypted and decrypted org users
- **migrate / 11-move-organization-keys-to-state-providers**:  Moves data from accounts.keys.organizationKeys to a top-level user-scoped location.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: libs/common/src/platform/services/key-state/org-keys.state.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  deserializer: (obj) => {
+    const result: Record<OrgId, OrgKey> = {};
+    for (const orgId of Object.keys(obj ?? {}).map((x) => x as OrgId)) {
+      result[orgId] = SymmetricCryptoKey.fromJSON(obj[orgId]) as OrgKey;
+    }
+    return result;
+  },
+  derive: async (from, { cryptoService }) => {
+    const result: Record<OrgId, OrgKey> = {};
+    for (const orgId of Object.keys(from ?? {}).map((x) => x as OrgId)) {
```
Comment (justindbaur) [inline libs/common/src/platform/services/key-state/org-keys.state.ts:29]: Why not just do a cast on the whole array like `Object.keys(from ?? {}) as OrgId[]`

File: libs/common/src/state-migrations/migrate.ts [hunk 2]
```diff
@@ -42,7 +43,9 @@ export async function migrate(
     .with(MoveBiometricAutoPromptToAccount, 6, 7)
     .with(MoveStateVersionMigrator, 7, 8)
     .with(MoveBrowserSettingsToGlobal, 8, 9)
-    .with(EverHadUserKeyMigrator, 9, CURRENT_VERSION)
+    .with(EverHadUserKeyMigrator, 9, 10)
+    .with(OrganizationKeyMigrator, 10, 11)
```
Comment (justindbaur) [inline libs/common/src/state-migrations/migrate.ts:47]:
```suggestion
    .with(OrganizationKeyMigrator, 10, CURRENT_VERSION)
```

File: libs/common/src/state-migrations/migrations/11-move-org-keys-to-state-providers.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+export class OrganizationKeyMigrator extends Migrator<10, 11> {
+  async migrate(helper: MigrationHelper): Promise<void> {
+    const accounts = await helper.getAccounts<ExpectedAccountType>();
+    async function migrateAccount(userId: string, account: ExpectedAccountType): Promise<void> {
+      const value = account?.keys?.organizationKeys?.encrypted;
+      if (value != null) {
+        await helper.setToUser(userId, USER_ENCRYPTED_ORGANIZATION_KEYS, value);
+        delete account.keys.organizationKeys;
+      }
+      await helper.set(userId, account);
```
Comment (justindbaur) [inline libs/common/src/state-migrations/migrations/11-move-org-keys-to-state-providers.ts:34]: I feel like this belongs in the if block. Unneeded operation if the value hasn't changed at all.

File: libs/common/src/platform/services/key-state/org-keys.state.ts (hunk 1 above)
Comment (MGibson1) [reply libs/common/src/platform/services/key-state/org-keys.state.ts:29]: 🤦  good point...

**Review by justindbaur — APPROVED** [review]

**Review by justindbaur — DISMISSED** [review]

**Review by shane-melton — APPROVED** [review]

**Review by r-tome — APPROVED** [review]

**Review by justindbaur — APPROVED** [review]

**Review by r-tome — APPROVED** [review]

**Review by shane-melton — APPROVED** [review]

**Pull request merged**